packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -31,6 +33,134 @@ typedef void _RegisterServiceExtensionCallback({
   @required ServiceExtensionCallback callback
 });
 
+/// A layer that mimics the behavior of another layer.
+///
+/// A proxy layer is used for cases where a layer needs to be added to a tree
+/// of layers but should not be removed from its existing tree of layers.
+class _ProxyLayer<L extends Layer> extends Layer {
+  final L _layer;
+
+  _ProxyLayer(this._layer);
+
+  @override
+  void addToScene(ui.SceneBuilder builder, Offset layerOffset) {
+    _layer.addToScene(builder, layerOffset);
+  }
+
+  @override
+  S find<S>(Offset regionOffset) => _layer.find(regionOffset);
+}
+
+/// A layer that renders like an [OffsetLayer] with the offset modified.
+///
+/// The [OffsetLayer] interface is not implemented as generally all methods
+/// would need to throw an exception to avoid accidentally modifying the layer
+/// being proxied.
+class _OffsetProxyLayer extends _ProxyLayer<OffsetLayer> {
+  final Offset _offset;
+
+  _OffsetProxyLayer(OffsetLayer _layer, this._offset) : super(_layer);
+
+  Offset _transformOffset(Offset offset) => offset + _offset - _layer.offset;
+
+  @override
+  void addToScene(ui.SceneBuilder builder, Offset layerOffset) {
+    _layer.addToScene(builder, _transformOffset(layerOffset));
+  }
+
+  @override
+  S find<S>(Offset regionOffset) {
+    return _layer.find(_transformOffset(regionOffset));
+  }
+}
+
+/// A place to paint to create screenshots of [RenderObject]s.
+///
+/// Requires that the render objects have already been painted successfully as
+/// part of the regular rendering pipeline.
+class _ScreenshotPaintingContext extends PaintingContext {
```

Comment:
i think this would be better as just `implements PaintingContext`. I'm not very comfortable with making `PaintingContext` extensible. There's all kinds of ways that can be misused in ways that would be really hard to help people disentangle. I'd much rather people just never tried.

Reply:
never mind